extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/impl/ReadMongoP.java

```diff
@@ -383,20 +400,29 @@ boolean everCompletes() {
         @Nonnull
         @Override
         public Traverser<?> nextChunkTraverser() {
-            Traverser<ChangeStreamDocument<Document>> traverser = cursor::tryNext;
-            return traverser.flatMap(doc -> {
-                resumeToken = doc.getResumeToken();
-                long eventTime = clusterTime(doc);
-                I item = mapFn.apply(doc);
-                return item == null
-                        ? Traversers.empty()
-                        : eventTimeMapper.flatMapEvent(item, 0, eventTime);
-            });
+            try {
+                List<ChangeStreamDocument<Document>> chunk = readChunk(cursor, BATCH_SIZE);
```

Comment:
Can't we really do it without the chunk?

Reply:
It's analogic to Kafka processor and it's in EventTimeMapper docs - https://github.com/hazelcast/hazelcast/blob/7c3894b1474dba56648cf10ad8441615486fcdea/hazelcast/src/main/java/com/hazelcast/jet/core/EventTimeMapper.java#L85

I tried to implement some Traverser that would do the same item-by-item, but I got:
```java
AssertionError: the traverser returned previously not yet drained: remove all items from the traverser before you call this method again.
	at com.hazelcast.jet.core.EventTimeMapper.flatMapEvent(EventTimeMapper.java:215) ~[classes/:?]
```

Reply:
Alternatively whole method may look like:
```java
 @Nonnull
        @Override
        public Traverser<?> nextChunkTraverser() {
            try {
                ChangeStreamDocument<Document> doc = cursor.tryNext();
                if (doc == null) {
                    return eventTimeMapper.flatMapIdle();
                } else {
                    long eventTime = clusterTime(doc);
                    I item = mapFn.apply(doc);
                    return eventTimeMapper.flatMapEvent(item, 0, eventTime);
                }
            } catch (MongoException e) {
                throw new JetException("error while reading from mongodb", e);
            }
        }

        private long clusterTime(ChangeStreamDocument<Document> changeStreamDocument) {
            BsonDateTime time = changeStreamDocument.getWallTime();
            return time == null ? System.currentTimeMillis() : time.getValue();
        }
```

Upside: no sub-batching
Downside: possible latency hit; will now deal with one event at each `complete()` invocation

Reply:
Let's merge it as is if it fixes the issue. It would be better without the buffer, but I couldn't make get the resiliency test to work. We really need some _Everything you wanted to know about watermarks_ tutorial.